### ISSUE TITLE
[no-unused-vars] does not count JSX as usage

### ISSUE THREAD

**Issue by borovik96** [typescript-eslint/typescript-eslint#111]
**Repro**

```JSON
{
  "extends": ["plugin:@typescript-eslint/recommended"],
  "plugins": ["@typescript-eslint/eslint-plugin"],
  "rules": {
    "react/jsx-filename-extension": [1, {
      "extensions": [".ts", ".tsx", ".js", ".jsx"]
    }],
  },
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2018,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    }
  },
}
```

```TS
import React from 'react'; // 'React' is defined but never used. [@typescript-eslint/no-unused-vars]
import Document from 'next/document';
import {
  ServerStyleSheet
} from 'styled-components';

export default class MyDocument extends Document {
  public static async getInitialProps(ctx): DefaultDocumentIProps {
    const sheet = new ServerStyleSheet();

    const originalRenderPage = ctx.renderPage;
    ctx.renderPage = () => originalRenderPage({
      enhanceApp: App => props => sheet.collectStyles(<App {...props} />) // 'App' is defined but never used. [@typescript-eslint/no-unused-vars]
    });

    const initialProps = await Document.getInitialProps(ctx);
    return {
      ...initialProps,
      styles: [...initialProps.styles, ...sheet.getStyleElement()]
    };
  }
}

```

**Expected Result**
Rule not trigger if a variable used as jsx
**Actual Result**
Rule trigger if a variable used as jsx
**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.0.0` |
| `@typescript-eslint/parser`        | `1.0.0` |
| `TypeScript`                       | `^3.2.4` |
| `ESLint`                           | `^5.12.1` |
| `node`                             | `v10.11.0` |
| `npm`                              | `6.4.1` |

**Comment by j-f1** [comment]
We’ve been asking people to use the [`react/jsx-uses-vars`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-uses-vars.md) rule, but I’m wondering if we should include this as a default behavior (cc @typescript-eslint/core-team)

**Comment by JamesHenry** [comment]
No, I don't think we would want to build in special consideration for react. This is the reason why ESLint changed its behaviour towards it: https://eslint.org/blog/2015/03/eslint-0.17.0-released#changes-to-jsxreact-handling

We can definitely provide documentation around this, though!

**Comment by borovik96** [comment]
I added `"plugin:react/recommended"` to extedns array and it works now. I can add this info to doc for [no-unused-vars] rule

**Comment by bradzacher** [comment]
Because we lost the context in the migration, issue on previous repo:
https://github.com/bradzacher/eslint-plugin-typescript/issues/283

I'll close this issue as the bug itself is working as intended.
Feel free to open a PR to add to the docs if you want!

**Comment by bradzacher** [comment]
@Jessidhia - this specific issue is about `no-unused-vars` not recognising usage of `import React`, or JSX component usage (`const X = ....; <X />`).

Please open a separate issue for the bug you have found, as it is instead about variable usage in JSX attributes (which I believe should be covered by our plugin?)

**Comment by TidyIQ** [comment]
So I just installed this and I'm not sure how to use it properly. Exactly what should I add to my config to stop this from happening in my React app?

**Comment by bradzacher** [comment]
As mentioned in the first reply to the issue.

> Use the [`react/jsx-uses-vars`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-uses-vars.md) rule.

There is documentation there for setting up the plugin.

**Comment by TidyIQ** [comment]
It doesn't mention anything about the config file on that page. Do I need to install a separate plugin?

**Comment by bradzacher** [comment]
Hop into the root of that repo and there should be more information about the linked plugin... 

You need to install `eslint-plugin-react`

**Comment by alexanderby** [comment]
For those who uses JSX without React, here is a solution:

1. Install [`eslint-plugin-local`](https://www.npmjs.com/package/eslint-plugin-local) package.
2. Create `.eslintplugin.js` file in project root.
3. Declare some custom rules (based on `eslint-plugin-react`):
```javascript
module.exports.rules = {
    'jsx-uses-my-pragma': {
        create(context) {
            const pragma = 'my';
            const usePragma = () => context.markVariableAsUsed(pragma);
            return {
                JSXOpeningElement: usePragma,
                JSXOpeningFragment: usePragma,
            };
        },
    },
    'jsx-uses-vars': {
        create(context) {
            return {
                JSXOpeningElement(node) {
                    if (node.name.name) {
                        const variable = node.name.name;
                        context.markVariableAsUsed(variable);
                    }
                },
            };
        },
    },
};
```
4. Add declared rules to ESLint config:
```javascript
module.exports = {
    parser: '@typescript-eslint/parser',
    plugins: ['@typescript-eslint', 'local'],
    extends: ['plugin:@typescript-eslint/recommended'],
    rules: {
        // ...
    },
    overrides: [
        {
            files: ['**/*.tsx'],
            rules: {
                'local/jsx-uses-my-pragma': 'error',
                'local/jsx-uses-vars': 'error',
            },
        },
    ],
};
```
5. Run `eslint`, no more `no-unused-vars`!

**Comment by AhmedBHameed** [comment]
> I added `"plugin:react/recommended"` to extedns array and it works now. I can add this info to doc for [no-unused-vars] rule

That solved my issue thanks @borovik96
